File: tests/unit/test_control_nodes.py
"""Compare/If are typed expressions; both upstream branches remain emitted."""
import copy
import json
from pathlib import Path
import subprocess
import unittest
import sgrape_core as c


def control_graph(nodes, edges=(), target='top', stage='pixel'):
    graph=c.demo_graph('color',target)
    graph['declarations']=[]
    graph['stages'][stage]={'nodes':[*nodes,c.node(stage+'_out','result')],
                            'edges':[*edges,c.edge(nodes[-1]['id'],'result','color' if stage=='pixel' else 'position')]}
    return graph


class ControlNodes(unittest.TestCase):
    def test_compare_operators_scalar_types_and_shader_stages(self):
        for target,stage in [('top','pixel'),('mat','pixel'),('mat','vertex')]:
            for ty in c.COMPARE_TYPES:
                for operator in c.COMPARE_OPERATORS:
                    with self.subTest(target=target,stage=stage,type=ty,operator=operator):
                        compare=c.node('compare','compare',type=ty,operator=operator)
                        compare['inputValues']={'a':3,'b':2}
                        graph=control_graph([compare],target=target,stage=stage)
                        before=copy.deepcopy(graph);result=c.compile_graph(graph)
                        self.assertEqual(graph,before)
                        self.assertIn('const bool sg_n_compare = ('+c.literal(3,ty)+' '+operator+' '+c.literal(2,ty)+');',result[stage])
                        self.assertEqual(result['stages'][stage]['ports']['compare'],{'in':{'a':ty,'b':ty},'out':{'out':'bool'}})
                        self.assertEqual(result,c.compile_graph(json.loads(json.dumps(graph))))

    def test_if_unconnected_defaults_and_all_numeric_result_types(self):
        for ty in c.TYPES:
            branch=c.node('if','branch',type=ty)
            reduction=c.node('length','length',type=ty)
            graph=control_graph([branch,reduction],[c.edge('branch','length','value')])
            code=c.compile_graph(graph)['pixel']
            self.assertIn('const '+ty+' sg_n_branch = (false ? '+c.literal(c.filled_value(ty,1),ty)+' : '+c.literal(c.filled_value(ty,0),ty)+');',code)
            branch['inputValues']={'condition':True,'true':c.filled_value(ty,.25),'false':c.filled_value(ty,.75)}
            self.assertIn('(true ? '+c.literal(branch['inputValues']['true'],ty)+' : '+c.literal(branch['inputValues']['false'],ty)+')',c.compile_graph(graph)['pixel'])

    def test_compare_invalid_operator_type_and_input_are_rejected(self):
        for operator in ['===','=',') || true || (',None,False,[]]:
            graph=control_graph([c.node('compare','compare',operator=operator)])
            with self.subTest(operator=operator),self.assertRaisesRegex(c.GraphError,'Compare: choose') as caught:c.compile_graph(graph)
            self.assertEqual(caught.exception.node,'compare')
        for ty in ['vec2','vec3','vec4','bool','sampler2D']:
            with self.subTest(type=ty),self.assertRaises(c.GraphError):c.compile_graph(control_graph([c.node('compare','compare',type=ty)]))
        for source in [c.node('compare','source'),c.node('vec2','source')]:
            graph=control_graph([source,c.node('compare','compare')],[c.edge('source','compare','a')])
            with self.assertRaisesRegex(c.GraphError,'Compare accepts'):c.compile_graph(graph)
        for ty,bad in [('int',.5),('uint',-1),('int',2147483648),('uint',4294967296)]:
            compare=c.node('compare','compare',type=ty);compare['inputValues']={'a':bad}
            with self.subTest(type=ty,value=bad),self.assertRaises(c.GraphError):c.compile_graph(control_graph([compare]))

    def test_if_condition_is_strict_bool_and_result_does_not_silently_truncate(self):
        graph=control_graph([c.node('float','number'),c.node('if','branch')],[c.edge('number','branch','condition')])
        with self.assertRaisesRegex(c.GraphError,'float cannot connect to bool'):c.compile_graph(graph)
        graph['stages']['pixel']['edges']=[c.edge('branch','result','color')]
        graph['stages']['pixel']['nodes'][1]['inputValues']={'condition':1}
        with self.assertRaisesRegex(c.GraphError,'boolean constant'):c.compile_graph(graph)
        graph=control_graph([c.node('vec3','value'),c.node('if','branch',type='vec2')],[c.edge('value','branch','true')])
        with self.assertRaisesRegex(c.GraphError,'vec3 cannot connect to vec2'):c.compile_graph(graph)
        for ty in ['int','uint','bool','sampler2D']:
            with self.subTest(type=ty),self.assertRaises(c.GraphError):c.compile_graph(control_graph([c.node('if','branch',type=ty)]))

    def test_runtime_and_specialization_conditions_keep_both_dependencies(self):
        for kind,ty,value in [('uniform','float',.5),('spec_constant','int',0),('spec_constant','uint',4294967295)]:
            condition=c.node(kind,'mode',declarationId='mode')
            compare=c.node('compare','compare',type=ty,operator='==');compare['inputValues']={'b':0}
            branch=c.node('if','branch')
            graph=control_graph([condition,compare,c.node('sin','true_path'),c.node('cos','false_path'),branch],
                    [c.edge('mode','compare','a'),c.edge('compare','branch','condition'),c.edge('true_path','branch','true'),c.edge('false_path','branch','false')])
            graph['declarations']=[dict(id='mode',kind=kind,name='uMode',type=ty,value=value,**({'constantId':7} if kind=='spec_constant' else {}))]
            code=c.compile_graph(graph)['pixel']
            self.assertIn('bool sg_n_compare = (',code);self.assertNotIn('const bool sg_n_compare',code)
            self.assertIn('float sg_n_branch = (sg_n_compare ? sg_n_true_path : sg_n_false_path);',code)
            self.assertLess(code.index('sg_n_true_path ='),code.index('sg_n_branch ='));self.assertLess(code.index('sg_n_false_path ='),code.index('sg_n_branch ='))
            if kind=='spec_constant':
                self.assertIn('layout(constant_id = 7) const '+ty+' uMode',code)
                graph['declarations'][0]['value']=1
                changed=c.compile_graph(graph)['pixel']
                self.assertIn('(sg_n_compare ? sg_n_true_path : sg_n_false_path)',changed)

    def test_literal_condition_does_not_prune_runtime_branch_or_binding(self):
        branch=c.node('if','branch');branch['inputValues']={'condition':False}
        graph=control_graph([c.node('uniform','unused_at_runtime',declarationId='value'),branch],[c.edge('unused_at_runtime','branch','true')])
        graph['declarations']=[dict(id='value',kind='uniform',name='uStillNeeded',type='float',value=.75)]
        code=c.compile_graph(graph)['pixel']
        self.assertIn('uniform float uStillNeeded;',code)
        self.assertIn('(false ? sg_n_unused_at_runtime : 0.0)',code)
        self.assertNotIn('const float sg_n_branch',code)
        branch['params']['requireConstant']=True
        with self.assertRaisesRegex(c.GraphError,'Require Constant'):c.compile_graph(graph)

    def test_compare_if_constant_requirements_and_explicit_branch_casts(self):
        compare=c.node('compare','condition',requireConstant=True)
        branch=c.node('if','branch',requireConstant=True,type='vec4')
        graph=control_graph([compare,c.node('float','value'),branch],[c.edge('condition','branch','condition'),c.edge('condition','branch','false'),c.edge('value','branch','true')])
        code=c.compile_graph(graph)['pixel']
        self.assertIn('const bool sg_n_condition',code)
        self.assertIn('const vec4 sg_n_branch = (sg_n_condition ? vec4(sg_n_value) : vec4(sg_n_condition));',code)
        for ty,value in [('int',2147483647),('uint',4294967295),('bool',True)]:
            branch=c.node('if','branch')
            graph=control_graph([c.node('spec_constant','source',declarationId='source'),branch],[c.edge('source','branch','true')])
            graph['declarations']=[dict(id='source',kind='spec_constant',name='sValue',type=ty,value=value,constantId=2)]
            self.assertIn('(false ? float(sValue) : 0.0)',c.compile_graph(graph)['pixel'])

    def test_bool_subgraph_interface_can_feed_if(self):
        graph=control_graph([{'id':'call','definitionUuid':c.CALL,'params':{'functionId':'choose'}}])
        graph['functions']=[dict(id='choose',name='Choose',scope='local',stages=['pixel'],
            inputs=[dict(id='condition',name='Condition',type='bool',default=True)],outputs=[dict(id='out',name='Out',type='float',default=0)],
            graph=dict(nodes=[{'id':'input','definitionUuid':c.FUNCTION_INPUT,'params':{}},c.node('if','branch'),{'id':'output','definitionUuid':c.FUNCTION_OUTPUT,'params':{}}],
                       edges=[c.edge('input','branch','condition','condition'),c.edge('branch','output','out')]))]
        self.assertIn(' ? 1.0 : 0.0)',c.compile_graph(graph)['pixel'])

    def test_javascript_type_plans_defaults_and_transaction_roundtrips(self):
        root=Path(__file__).resolve().parents[2]
        payload={'catalog':list(c.CATALOG.values()),'contract':c.type_contract()}
        result=subprocess.run(['node',str(root/'tests/unit/test_control_nodes.js')],input=json.dumps(payload),text=True,capture_output=True)
        self.assertEqual(result.returncode,0,result.stderr)
        for graph in json.loads(result.stdout)['graphs']:c.compile_graph(graph)


if __name__=='__main__':unittest.main()

File: tests/unit/test_control_nodes.js
/* Real connection plans and edit transactions; only DOM/network work is stubbed. */
const assert=require('node:assert/strict'),fs=require('node:fs'),vm=require('node:vm'),path=require('node:path');
const payload=JSON.parse(fs.readFileSync(0,'utf8')),dir=path.resolve(__dirname,'../../src/editor');
const elements=new Map(),element=key=>{if(!elements.has(key))elements.set(key,{value:'all',textContent:'',title:'',hidden:false,disabled:false,focus(){},replaceChildren(){},setAttribute(){},addEventListener(){},classList:{add(){},remove(){},toggle(){}}});return elements.get(key);};
const context=vm.createContext({assert,payload,console,TextEncoder,crypto:globalThis.crypto,
  location:{pathname:'/',hash:''},history:{replaceState(){}},window:{addEventListener(){}},
  document:{addEventListener(){},querySelector:element,querySelectorAll:()=>[]},
  sessionStorage:{getItem(){return '';},setItem(){}},setTimeout(){return 1;},clearTimeout(){}});
for(const name of ['functions_model.js','functions_ui.js','graph_ui.js','inspector.js'])vm.runInContext(fs.readFileSync(path.join(dir,name),'utf8'),context);
const app=fs.readFileSync(path.join(dir,'app.js'),'utf8');
vm.runInContext(app.slice(0,app.indexOf("$('#canvas').addEventListener('dragover'")),context);
vm.runInContext(`
render=()=>{};wires=()=>{};renderGraphEditActions=()=>{};inspector=()=>{};renderNavigation=()=>{};renderNativeSourceValues=()=>{};refreshUniforms=()=>{};
catalog=payload.catalog;setTypeContract(payload.contract);editorTarget='top';
const graphs=[];
function node(key,id,params={}){const d=catalog.find(d=>d.key===key);return {id,definitionUuid:d.definitionUuid,revisionHash:d.revisionHash,params:{...clone(d.defaults),...params},ui:{x:24,y:24,...(supportsAutoType(d)?{typeMode:'auto'}:{})}};}
const edge=(from,to,port,out='out')=>({from:[from,out],to:[to,port]});
function setup(nodes,edges=[]){graph={schemaVersion:1,target:'top',topSourceVersion:1,topInputs:[],declarations:[],functions:[],stages:{pixel:{nodes:[...nodes,node('pixel_out','result')],edges}}};graphTrail=[];past=[];future=[];selected=null;selection.clear();selectedEdge=null;dirty=false;}
const n=id=>current().nodes.find(n=>n.id===id),d=key=>catalog.find(d=>d.key===key);
const info=(id,kind,port)=>({node:id,kind,port,type:ports(n(id),kind)[port]});
const connect=(a,b,p,o='out')=>connectPorts(info(a,'outputs',o),info(b,'inputs',p));
const identical=(a,b)=>assert.equal(JSON.stringify(a),JSON.stringify(b));
function rejectsWire(a,b,p,o='out'){const before=clone({graph,past,future,dirty});assert.equal(connect(a,b,p,o),false);identical({graph,past,future,dirty},before);}

assert.equal(supportsAutoType(d('compare')),true);assert.equal(supportsAutoType(d('if')),true);
// Matrix/double operation overloads follow the additive value-type batch.
const legacyValueTypes=valueTypes().filter(type=>typeFamily(type)!=='double'&&!isMatrixType(type));
identical(selectableNodeTypes(d('compare')),['float','int','uint']);identical(selectableNodeTypes(d('if')),legacyValueTypes);
for(const ty of selectableNodeTypes(d('compare'))){const compare=node('compare','compare',{type:ty});assert.equal(defaultInput(compare,'a',ty),0);assert.equal(defaultInput(compare,'b',ty),0);identical(resolvedNodePorts(d('compare'),compare.params,null,'outputs'),{out:'bool'});}
for(const ty of legacyValueTypes){const branch=node('if','branch',{type:ty});assert.equal(defaultInput(branch,'condition','bool'),false);identical(defaultInput(branch,'true',ty),filledValue(ty,1));identical(defaultInput(branch,'false',ty),filledValue(ty,0));}

// Only an unconnected Auto Compare uses int. Locked/legacy float and the
// remaining Auto operations retain their existing type/default policies.
setup([node('compare','auto'),node('compare','locked'),node('compare','legacy'),node('add','math'),node('if','branch')]);n('locked').ui.typeMode='locked';delete n('legacy').ui.typeMode;
assert.equal(change(()=>current().nodes.push(node('float','value'))),true);
assert.equal(n('auto').params.type,'int');assert.equal(n('locked').params.type,'float');assert.equal(n('legacy').params.type,'float');assert.equal(n('math').params.type,'float');assert.equal(n('branch').params.type,'float');
assert.equal(connect('value','auto','a'),true);assert.equal(n('auto').params.type,'float');
assert.equal(change(()=>setNodeInputValue(n('auto'),'b',1.75)),true);
const floatConnected=clone(graph);assert.equal(change(()=>current().edges=current().edges.filter(e=>e.to[0]!=='auto')),true);
assert.equal(n('auto').params.type,'int');assert.equal(n('auto').inputValues.b,1);
const intDisconnected=clone(graph);undo();identical(graph,floatConnected);undo(true);identical(graph,intDisconnected);
assert.equal(setMathType(n('auto'),'float'),true);assert.equal(n('auto').inputValues.b,1.75);assert.equal(n('auto').ui.typeMode,'locked');
assert.equal(setMathType(n('auto'),'auto'),true);assert.equal(n('auto').params.type,'int');
assert.equal(connect('value','auto','a'),true);assert.equal(n('auto').params.type,'float');assert.equal(n('auto').inputValues.b,1.75);

// Integer sources preserve exact integer comparison, while defaults retain a
// separate cache for each scalar type across automatic promotion/demotion.
for(const ty of ['int','uint']){
  setup([node('spec_constant','source',{declarationId:'source'}),node('compare','compare')]);
  graph.declarations=[{id:'source',kind:'spec_constant',name:'sValue',type:ty,value:2,constantId:0}];
  n('compare').inputValues={a:.75,b:-2.25};const before=clone(graph);
  assert.equal(connect('source','compare','a'),true);assert.equal(n('compare').params.type,ty);assert.equal(n('compare').inputValues.a,0);assert.equal(n('compare').inputValues.b,ty==='int'?-2:0);assert.equal(past.length,1);
  const typed=clone(graph);undo();identical(graph,before);undo(true);identical(graph,typed);
  current().nodes.push(node('convert','boolCast',{fromType:'bool',toType:'vec4'}));assert.equal(connect('compare','boolCast','value'),true);assert.equal(connect('boolCast','result','color'),true);graphs.push(clone(graph));
  assert.equal(change(()=>current().edges=current().edges.filter(e=>e.to[0]!=='compare')),true);assert.equal(n('compare').params.type,'int');
  assert.equal(setMathType(n('compare'),'float'),true);identical(n('compare').inputValues,{a:.75,b:-2.25});assert.equal(setMathType(n('compare'),'auto'),true);
  assert.equal(connect('source','compare','a'),true);assert.equal(n('compare').params.type,ty);assert.equal(n('compare').inputValues.b,ty==='int'?-2:0);
}

// Auto prioritizes an exact input type; mixed signed/unsigned inputs need one
// cast rather than converting both sides to float. A locked type stays available.
setup([node('spec_constant','signed',{declarationId:'signed'}),node('spec_constant','unsigned',{declarationId:'unsigned'}),node('compare','compare')]);
graph.declarations=[{id:'signed',kind:'spec_constant',name:'sSigned',type:'int',value:-1,constantId:0},{id:'unsigned',kind:'spec_constant',name:'sUnsigned',type:'uint',value:4294967295,constantId:1}];
assert.equal(connect('signed','compare','a'),true);assert.equal(n('compare').params.type,'int');assert.equal(connect('unsigned','compare','b'),true);assert.equal(n('compare').params.type,'int');assert.equal(inputTypeDisplay(n('compare'),'a').conversion,null);assert.equal(inputTypeDisplay(n('compare'),'b').conversion,'cast');current().nodes.push(node('convert','boolCast',{fromType:'bool',toType:'vec4'}));assert.equal(connect('compare','boolCast','value'),true);assert.equal(connect('boolCast','result','color'),true);graphs.push(clone(graph));

// Mixed float/integer input order keeps the pre-existing score/tie-break rules.
for(const first of ['float','int','uint'])for(const second of ['float','int','uint']){
  setup([node('scalar','a',{type:first,value:2}),node('scalar','b',{type:second,value:3}),node('compare','compare')]);
  assert.equal(connect('a','compare','a'),true);assert.equal(n('compare').params.type,first);
  assert.equal(connect('b','compare','b'),true);assert.equal(n('compare').params.type,['float','int','uint'].find(type=>type===first||type===second));
}

setup([node('compare','condition'),node('if','branch'),node('vec3','vector',{value:[.1,.2,.3]}),node('float','scalar'),node('length','length')]);
assert.equal(connect('condition','branch','condition'),true);assert.equal(n('branch').params.type,'float');assert.equal(connect('vector','branch','true'),true);assert.equal(connect('scalar','branch','false'),true);assert.equal(n('branch').params.type,'vec3');
assert.equal(inputTypeDisplay(n('branch'),'false').conversion,'splat');assert.equal(connect('branch','length','value'),true);assert.equal(connect('length','result','color'),true);graphs.push(clone(graph));
const conditionRevision=n('condition').revisionHash;
for(const operator of ['>','>=','<','<=','==','!=']){const previous=clone(graph),count=past.length;if(n('condition').params.operator===operator)continue;assert.equal(change(()=>n('condition').params.operator=operator),true);assert.equal(past.length,count+1);assert.equal(n('condition').revisionHash,conditionRevision);const changed=clone(graph);undo();identical(graph,previous);undo(true);identical(graph,changed);graphs.push(clone(graph));}
rejectsWire('scalar','branch','condition');rejectsWire('vector','condition','a');rejectsWire('condition','condition','a');
assert.equal(change(()=>current().nodes.push(node('compare','other'))),true);rejectsWire('condition','other','a');
assert.throws(()=>creatorTypePlan(d('compare'),typeVariants(d('compare'))[0],'a',info('condition','outputs','out'),false));
assert.equal(change(()=>current().nodes.push(node('vec2','pair'))),true);rejectsWire('pair','branch','false');

// Boolean-to-numeric conversion is always explicit; a numeric value cannot
// replace the strict Boolean condition.
rejectsWire('condition','branch','false');current().nodes.push(node('convert','boolCast',{fromType:'bool',toType:'vec3'}));assert.equal(connect('condition','boolCast','value'),true);assert.equal(connect('boolCast','branch','false'),true);assert.equal(n('branch').params.type,'vec3');graphs.push(clone(graph));
setup([node('compare','condition'),node('if','branch',{type:'vec4'}),node('float','scalar')]);n('branch').ui.typeMode='locked';
assert.equal(connect('condition','branch','condition'),true);assert.equal(connect('scalar','branch','true'),true);assert.equal(n('branch').params.type,'vec4');assert.equal(connect('branch','result','color'),true);graphs.push(clone(graph));

// Graph JSON and clipboard use the regular immutable catalog identities.
const exported=JSON.parse(JSON.stringify(graph));identical(exported,graph);const nodeBefore=clone(n('branch'));
const data=GraphClipboard.decode(GraphClipboard.encode(graph,current(),['branch'],shaderId));
assert.equal(change(()=>GraphClipboard.paste(graph,current(),data,{source:shaderId,stage,target:editorTarget,catalog,types:numericTypes(),anchor:{x:400,y:300}})),true);
const duplicate=current().nodes.find(item=>item.definitionUuid===nodeBefore.definitionUuid&&item.id!==nodeBefore.id);assert.equal(duplicate.revisionHash,nodeBefore.revisionHash);identical(duplicate.params,nodeBefore.params);assert.equal(duplicate.ui.typeMode,'locked');graphs.push(clone(graph));
console.log(JSON.stringify({passed:true,graphs}));
`,context);
